Add Explanation of some Utilities to User Guide (#392)
emdgroup/baybe: docs/userguide/utils.md

scenarios.

## Search Space Memory Estimation
In search spaces that have discrete parts, the memory needed to store the respective
data can become excessively large as the number of points grows with the amount of
possible combinations arising form all discrete parameter values.

The [`SearchSpace.estimate_product_space_size`](baybe.searchspace.core.SearchSpace.estimate_product_space_size)
and [`SubspaceDiscrete.estimate_product_space_size`](baybe.searchspace.discrete.SubspaceDiscrete.estimate_product_space_size)
utilities allows estimating the memory needed to represent the discrete subspace. 
It will return a [`MemorySize`](baybe.searchspace.discrete.MemorySize) object that
contains some relevant estimates.

```python
import numpy as np

from baybe.parameters import NumericalDiscreteParameter
from baybe.searchspace import SearchSpace

# This will create 10 parameters with 20 values each
# The resulting space would have 20^10 entries, requiring around 745 TB of memory for
# both experimental and computational representation of the search space
parameters = [
    NumericalDiscreteParameter(name=f"p{k+1}", values=np.linspace(0, 100, 20))
    for k in range(10)
]

> # Estimate the required memory for such a space in Bytes

Author: true, is leftover, removed

mem_estimate = SearchSpace.estimate_product_space_size(parameters)

# Print quantities of interest
print("Experimental Representation")
print(f"Estimated size: {mem_estimate.exp_rep_human_readable}")
print(f"Estimated size in Bytes: {mem_estimate.exp_rep_bytes}")
print(f"Expected data frame shape: {mem_estimate.exp_rep_shape}")

print("Computational Representation")
print(f"Estimated size: {mem_estimate.comp_rep_human_readable}")
print(f"Estimated size in Bytes: {mem_estimate.comp_rep_bytes}")
print(f"Expected data frame shape: {mem_estimate.comp_rep_shape}")
```

```{admonition} Estimate with Constraints
:class: warning

> `estimate_product_space_size` currently does not include the influence of potential

Author: added

constraints in your search space as it is generally very hard to incorporate the effect
of arbitrary constraints without actually buidling the entire space. Hence, you should
always **treat the number you get as upper bound** of required memory. This can still be
useful - for instance if your estimate already is several Exabytes, it is unlikely that
most computers would be able to handle the result even if there are constraints present.
```

> ```{admonition} Influence of Continuous Parameters

Author: sorted

:class: info
Continuous parameters fo not influence the size of the discrete search space part.
Hence, they are ignored by the utility.
```

```{admonition} Memory During Optimization
:class: warning

> `estimate_product_space_size` only estimates the memory required to handle the search

Reviewer: again, link

Author: added

already during the Cartesian product, requiring substantially less memory overall. In
addition, BayBE can also be installed with an optional `polars` dependency (`pip install
baybe[polars]`) that will utilize the more efficient machinery form polars for handling
of the search space and its constraints.
```

## Reproducibility
In some scenarios, for instance when testing your code setup, it can be useful to fix
the random seeds for all relevant engines to generate reproducible results. BayBE offers
the [`set_random_seed`](baybe.utils.random.set_random_seed) utility for this purpose:

```python
from baybe.utils.random import set_random_seed

# Set the global random seed for all relevant engines
set_random_seed(1337)

# Assuming we have a prepared campaign
campaign.recommend(5)
```

Setting the global random seed can be undesirable if there are other packages in your
setup. For this, BayBE offers [`temporary_seed`](baybe.utils.random.temporary_seed):

```python
from baybe.utils.random import temporary_seed

# Set the random seed for all relevant engines temporarily within the context
with temporary_seed(1337):
    campaign.recommend(5)
```

> ## Add Fake Target Measurements and Noise

Reviewer: ```suggestion
## Adding Fake Target Measurements and Parameter Noise
```

> When creating test scripts, it is often useful to try the recommendation loop for a few

Reviewer: I think this might now be the time to finally harmonize some of our terminology. Perhaps we should find a more consistent/precise name for `add_fake_results`, just like we have done in other places when we talk about `measurements` etc. So what is the most accurate wording here, since `results` is used nowhere else and is very generic? `add_fake_measurements`? `add_fake_targets_values`? Alternatively, we could also change it to something like `fantasize_measurements` or similar.

Author: renamed to `add_fake_measurements`
no deprecations for utilities

iterations. However, this requires some arbitrary target measurements to be set. Instead
of coming up with a custom logic every time, you can use the
[`add_fake_results`](baybe.utils.dataframe.add_fake_results) utility to add fake target
measurements and the [`add_parameter_noise`](baybe.utils.dataframe.add_parameter_noise)
utility to add artificial parameter noise:

```python

> from baybe.utils.dataframe import add_fake_results, add_parameter_noise

Reviewer: One other thing: do we want to finally fix the inplace behavior (i.e. have we already talked about that?). I think we should either make the inplace an optional flag or at least always also return the modified dataframe. Thoughts welcome

Author: so remind me again, this was achieved by modifying inplace + returning the modified object right? Dont see much reason against that and can add it

Author: changed

# Get recommendations
recommendations = campaign.recommend(5)

> # Add fake target measurements and artificial parameter noise to the recommendations
> # The utilities will modify the data frames inplace

Reviewer: ```suggestion
# Add fake target measurements and artificial parameter noise to the recommendations.
# The utilities modify the dataframes inplace.
```

measurements = recommendations.copy()
add_fake_results(measurements, campaign.targets)
add_parameter_noise(measurements, campaign.parameters)

> # Now continue the loop, e.g by adding the measurements...

Reviewer: ```suggestion
# Now continue the loop, e.g. by adding the measurements ...
```

Author: is the second space somethign im not aware of? like emdash? seems strange to me triple period imo has no space before

print(f"Estimated size in Bytes: {mem_estimate.comp_rep_bytes}")
print(f"Expected data frame shape: {mem_estimate.comp_rep_shape}")
```

```{admonition} Estimation with Constraints
:class: warning
{meth}`~baybe.searchspace.core.SearchSpace.estimate_product_space_size`
currently does not include the influence of potential constraints in your search space
as it is generally very hard to incorporate the effect of arbitrary constraints without
actually building the entire space. Hence, you should always **treat the number you get
as upper bound** of required memory. This can still be useful – for instance if your
estimate already is several Exabytes, it is unlikely that most computers would be able
to handle the result even if there are constraints present.
```

```{admonition} Memory During Optimization
:class: warning
{meth}`~baybe.searchspace.core.SearchSpace.estimate_product_space_size`
only estimates the memory required to handle the search space. **It does not estimate
the memory required during optimization**, which can be of a similar magnitude, but
generally depends on additional factors.
```

```{admonition} Influence of Continuous Parameters
:class: info

> Continuous parameters do not influence the size of the discrete search space part.

Reviewer 2: Maybe add a half-sentence about the reason (e.g. "as they do not contribute to the combinatorial explosion of creating the configurations of discrete parameters" or similar)?

Author: added

Hence, they are ignored by the utility.
```

```{admonition} Efficient Search Space Creation
:class: tip
If you run into issues creating large search spaces, as for instance in mixture
use cases, you should consider resorting to more specialized ways of creation by
invoking alternative search space constructors like 
{meth}`~baybe.searchspace.discrete.SubspaceDiscrete.from_dataframe`
or 
{meth}`~baybe.searchspace.discrete.SubspaceDiscrete.from_simplex`.
Instead of creating a product space first and then filtering it down
according to constraints, they offer a more direct and thus efficient path to the 
desired result, typically requiring substantially less memory. 
For example, {meth}`~baybe.searchspace.discrete.SubspaceDiscrete.from_simplex` 
includes the mixture constraint already *during* the product creation. 
In addition, BayBE can also be installed with its optional `polars` dependency 

> (`pip install baybe[polars]`) that activates efficient machinery for constraint handling.

Reviewer 2: Polars cannot be installed in computers with chips that do not have AVX support, see https://docs.pola.rs/user-guide/installation/#legacy-cpu. Since the M1 Chip is one of the chips affected by this, we might want to add a comment here that it might be necessary to manually install the "legacy" package. Not sure if this is the correct place for this information, however.

Author: please mark this to known issues or something, i dont think this admonition (which is essentially just a sidenote) is the place for this detail

Reviewer: @AVHopp, can you create an issue that reminds us of adding it to known_issues?

```

## Reproducibility
In some scenarios, for instance when testing your code setup, it can be useful to fix
the random seeds for all relevant engines to generate reproducible results. BayBE offers
the [`set_random_seed`](baybe.utils.random.set_random_seed) utility for this purpose:

```python
from baybe.utils.random import set_random_seed

# Set the global random seed for all relevant engines
set_random_seed(1337)

# Assuming we have a prepared campaign
campaign.recommend(5)
```

Setting the global random seed can be undesirable if there are other packages in your

> setup. For this, BayBE offers [`temporary_seed`](baybe.utils.random.temporary_seed):

Reviewer 2: "... that might unintentionally be influenced by this".

Author: added
